modyn/trainer_server/internal/dataset/online_dataset.py

```diff
@@ -138,25 +175,62 @@ def _info(self, msg: str, worker_id: Optional[int]) -> None:  # pragma: no cover
     def _debug(self, msg: str, worker_id: Optional[int]) -> None:  # pragma: no cover
         logger.debug(f"[Training {self._training_id}][PL {self._pipeline_id}][Worker {worker_id}] {msg}")
 
-    def _get_data(self, worker_id: int, partition_id: int) -> tuple[list[int], list[bytes], list[int]]:
-        self._info("Getting keys from selector", worker_id)
-        keys = self._get_keys_from_selector(worker_id, partition_id)
+    def switch_to_local_key_source(self) -> None:
+        self._key_source = KeySource.LOCAL
+        self._local_dataset_handler = LocalDatasetHandler(self._pipeline_id, self._trigger_id, self._number_of_workers)
+
+    def switch_to_selector_key_source(self) -> None:
+        if self._key_source == KeySource.SELECTOR:
+            return
+        self._key_source = KeySource.SELECTOR
+        assert self._local_dataset_handler is not None
+        self._local_dataset_handler.clean_working_directory()
+        self._local_dataset_handler = None
+
+    def _get_data(
+        self, worker_id: int, partition_id: int
+    ) -> tuple[list[int], list[bytes], list[int], Optional[list[float]]]:
+        # Keys can either be retreived from the selector or from local.
```

Comment:
Just an idea, but maybe we should rename local to local_downsampler? To make clear where the keys are actually coming from, they are supposed to come from the result of downsampling

Reply:
Renamed